pkg/apis/workspace/types.go

```diff
@@ -0,0 +1,75 @@
+package workspace
+
+const (
+	DefaultBlock         = "default"
+	ProjectSelectorField = "projectSelector"
+)
+
+// Workspace is a logical concept representing a target that stacks will be deployed to.
+// Workspace is managed by platform engineers, which contains a set of configurations
+// that application developers do not want or should not concern, and is reused by multiple
+// stacks belonging to different projects.
+type Workspace struct {
+	// Name identifies a Workspace uniquely.
+	Name string `yaml:"-" json:"-"`
+
+	// Modules are the configs of a set of modules.
+	Modules ModuleConfigs `yaml:"modules,omitempty" json:"modules,omitempty"`
+
+	// Runtimes are the configs of a set of runtimes.
+	Runtimes RuntimeConfigs `yaml:"runtimes,omitempty" json:"runtimes,omitempty"`
+
+	// Backends are the configs of a set of backends.
+	Backends BackendConfigs `yaml:"backends,omitempty" json:"backends,omitempty"`
+}
+
+// ModuleConfigs is a set of multiple ModuleConfig, whose key is the module name.
+type ModuleConfigs map[string]ModuleConfig
+
+// ModuleConfig is the config of a module, which contains a default and several patcher blocks.
+//
+// The default block's key is "default", and value is the module inputs. The patcher blocks' keys
+// are the patcher names, which are just block identifiers without specific meaning, but must
+// not be "default". Besides module inputs, patcher block's value also contains a field named
+// "projectSelector", whose value is a slice containing the project names that use the patcher
+// configs. A project can only be assigned in a patcher's "projectSelector" field, the assignment
+// in multiple patchers is not allowed. For a project, if not specified in the patcher block's
+// "projectSelector" field, the default config will be used.
+type ModuleConfig map[string]GenericConfig
+
+// RuntimeConfigs contains a set of runtime config.
+type RuntimeConfigs struct {
+	// Kubernetes contains the config to access a kubernetes cluster.
+	Kubernetes KubernetesConfig `yaml:"kubernetes,omitempty" json:"kubernetes,omitempty"`
```

Comment:
Why not design `KubernetesConfig` as a pointer since you need to check whether it is empty(nil)?